pkg/cloudprovider/drift.go

```diff
@@ -39,58 +39,51 @@ import (
 )
 
 const (
-	NodeClassDrift  cloudprovider.DriftReason = "NodeClassDrift"
-	K8sVersionDrift cloudprovider.DriftReason = "K8sVersionDrift"
-	ImageDrift      cloudprovider.DriftReason = "ImageDrift"
-	SubnetDrift     cloudprovider.DriftReason = "SubnetDrift"
+	NodeClassDrift       cloudprovider.DriftReason = "NodeClassDrift"
+	K8sVersionDrift      cloudprovider.DriftReason = "K8sVersionDrift"
+	ImageDrift           cloudprovider.DriftReason = "ImageDrift"
+	SubnetDrift          cloudprovider.DriftReason = "SubnetDrift"
+	KubeletIdentityDrift cloudprovider.DriftReason = "KubeletIdentityDrift"
 
 	// TODO (charliedmcb): Use this const across code and test locations which are signaling/checking for "no drift"
 	NoDrift cloudprovider.DriftReason = ""
 )
 
 func (c *CloudProvider) isNodeClassDrifted(ctx context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1beta1.AKSNodeClass) (cloudprovider.DriftReason, error) {
```

Comment:
Is there a case to be made to determine this before calling the funcs below (looks like not all of them need it but more than one does):

```
	node, err := c.getNodeForDrift(ctx, nodeClaim)
	if err != nil {
		return "", err
	}
	if node == nil {
		return "", nil
	}
```

The current flow is a bit weird in the sense that for the funcs that require the above to make a determination, we are kind of overloading an implicit ternary return semantic ("not enough input to actually determine drift") in the function itself. Maybe we hoist that out of the function and instead make it a precondition of calling the func at all, like a pre-step where we gather "which types are drift are we actually able to evaluate?"

Reply:
If we do end up wanting to do this "lazy at-most-once cacher" I don't think it makes sense to do it here because nodes are already cached by the controller-runtime client, the things we'd want to cache are like VMs or NICs or whatever, which today are only checked once so don't need to be cached.

Reply:
I wouldn't put this as a pre-check for all `is___Drifted` checks, but could see it being added before `isK8sVersionDrifted`, and `isKubeletIdentityDrifted`.

The reasoning I wouldn't put it before `areStaticFieldsDrifted` is that part of the current drift pattern is to short circuit on success, meaning that if we don't have to check for the `getNodeForDrift`, we might as well not.

My opinion is to leave it the way Matthew has it (I think its fine as is), but leave a TODO for possibly hoisting it out. Would only want it preformed after `areStaticFieldsDrifted` though.

Reply:
Actually, thinking on this more, I'm more mixed on it being better or not. I kind of like the individual cases doing the checks themselves, and leaving the easy loop over all cases Matthew has. 

The fact it is being used in two separate places does make it feel worth "reusing" and doing in one location, but as Matthew said with point 2 `OTOH, it's not expensive to do this check because it's reading the node from a cached client anyway.`, I think its fine? And extracting it would make the pass over things a bit more clunky keeping the priority ordering, and such.

Reply:
> lazy at-most-once cacher

- I think this is the solution I would lean towards, if we start finding this pattern creep up more, but agree its not necessary for nodes here.

Reply:
I've left a TODO comment here.

Reply:
all sgtm!